Source/WebCore/dom/Element.cpp

```diff
@@ -5363,4 +5374,43 @@ bool Element::isPopoverShowing() const
     return popoverData() && popoverData()->visibilityState() == PopoverVisibilityState::Showing;
 }
 
+// https://drafts.csswg.org/css-contain/#relevant-to-the-user
+bool Element::isRelevantToUser() const
+{
+    return hasRareData() && elementRareData()->contentRelevancyStatus();
+}
+
+OptionSet<ContentRelevancyStatus> Element::contentRelevancyStatus() const
+{
+    if (!hasRareData())
+        return { };
+    return elementRareData()->contentRelevancyStatus();
+}
+
+void Element::setContentRelevancyStatus(ContentRelevancyStatus status, bool value)
+{
+    if (contentRelevancyStatus().contains(status) != value) {
+        ensureElementRareData().setContentRelevancyStatus(status, value);
+        invalidateStyle();
+    }
+}
+
+void Element::updateContentRelevancyStatus(ContentRelevancyStatus status, bool value)
+{
+    if (!document().isObservingContentVisibilityTargets())
+        return;
+    for (auto target = this; target; target = target->parentElementInComposedTree()) {
+        if (target->renderer() && target->renderStyle()->contentVisibility() == ContentVisibility::Auto) {
+            target->setContentRelevancyStatus(status, value);
+            return;
```

Comment:
What if one of the offscreen ancestors has `content-visibility: auto;` ? Is it accurate to early return here in that case?

Reply:
For the non-nested situation, stopping at the first c-v: auto element is as intended.
For a nested c-v: auto situation, what I think you are describing, I am not sure the specification is strict there.
Anyway, I prefer to deal with this when adding support for ContentVisibilityAutoStateChangeEvent (https://bugs.webkit.org/show_bug.cgi?id=258880), where this is easier to test and existing should show up any problems or we can add new ones.
(FWIW right now removing the early return has no effect on the WPT test results).

Reply:
Would be good to make testcases with nested `content-visibility: auto` and see if we match other browsers here.

Reply:
I updated the PR and added content-visibility-auto-nested-scroll.html to test that. It turns out the early return has to be removed to match the other browsers for this test.